### PR TITLE
Add -flambda/-no-flambda compile-time options to ocamlopt

### PR DESCRIPTION
**Pull request by alainfrisch**
This PR is about allowing to compile with or without flambda independently of the default mode chosen at configure-time.

The configure-time option determines the default mode (and the one used to compiled stdlib and compiler-libs), but the restriction that code compiled with/without flambda (as determined at configure-time) cannot be linked together is lifted.

The same magic number is used in .cmx files for flambda and non-flambda mode.  Previously, different magic numbers were used, which leads to potentially cryptic error messages when trying to mix the two modes.

When compiling a unit B which imports A.cmx, in different mode w.r.t. flambda, cross-module optimization information from A.cmx are ignored (just as if A.cmx had been obtained by -opaque).  This also trigger a new warning 61 explaning the situation (instead of the above-mentioned error message).

In addition, this removes some code duplication between asmpackager.ml and compilenv.ml.

In the long term, if the non-flambda mode is maintained, one could hope to get more interoperability between the two modes, i.e. mapping cross-module info of one mode into the other one, in only one or both directions.  But even if we never reach that point, this PR will give more visibility to flambda since it simplifies testing flambdalocally (e.g. to see if some kinds of optimizations -- not involving external modules -- are triggered).  For quick testing, it is much simpler to pass "-flambda" than to rebuild a full OCaml configured with -flambda (which is rather slow).

There is also the case of very big and boring modules which takes a lot of time to compile even with -Oclassic, and for which performance is known to be completely irrelevant.  For these cases, one could use -no-flambda to speed up compilation.

There remains a TODO in Clflags (to initialize inlining threshold), hence the WIP in the title.

### CODE REVIEW

File: utils/config.mlp [hunk 1]
```diff
@@ -67,22 +67,14 @@ let mkdll, mkexe, mkmaindll =
...
-    "Caml1999Y016"
-  else
-    "Caml1999Y015"
-and cmxa_magic_number =
-  if flambda then
-    "Caml1999Z015"
-  else
-    "Caml1999Z014"
+and cmx_magic_number = "Caml1999Y015"
+and cmxa_magic_number = "Caml1999Z014"
```
Comment (chambart) [inline utils/config.mlp]: We probably don't want to decrease the maximum magic. Future compiler versions would risk reusing `Caml1999Y016`

**Comment by const-rs** [comment]
> Previously, different magic numbers were used, which leads to potentially cryptic error messages when trying to mix the two modes.

There are only two places which issue these messages - objinfo.ml and compilenv.ml. 

In compilenv.ml to make things clear one can just add "(or outdated format)" to error message at the very end of the file - Not_a_unit_info. This will cover both .cmx and .cmxa files. Of course format may be too new, but it will give a hint in most situations.

**Comment by mshinwell** [comment]
I would like to discuss this before any merge (I am on vacation at the moment).

**Comment by mshinwell** [comment]
(In particular this is probably related to #620.)

File: utils/config.mlp (hunk 1 above)
Comment (alainfrisch) [inline utils/config.mlp]: Good point, thanks.  I'm now using the magic numbers that used to be for the flambda case.

**Comment by alainfrisch** [comment]
> (In particular this is probably related to #620.)

Well, the two PRs could be combined so that a single compiler could be used for flambda and non-flambda projects by installing two versions of the stdlib.  That said, in practice, people are likely to use external packages (installed e.g. by OPAM), and I don't think we want to go to the direction where all installed packages would have to build and install multiple versions.

**Comment by alainfrisch** [comment]
@chambart, @mshinwell, @lpw25 : do you see big obstacles in allow a limited form of cross-module optimization between flambda and closure modes?  My guess is that supporting direct function calls, even if not inlining, should be rather simple, no?

**Comment by alainfrisch** [comment]
> I would like to discuss this before any merge (I am on vacation at the moment).

Ping @mshinwell.

**Comment by mshinwell** [comment]
Sorry, I failed to look at this one, and I'm about to be away again.

@alainfrisch To address your most recent comment, my general feeling is that we need to be looking to a world in which the old Closure code is removed.  To that end, given how stretched we are on manpower, I think it's a much better use of time to work on speed improvements to Flambda rather than trying to achieve interoperability.  I will let @lpw25 and @chambart comment.

**Comment by alainfrisch** [comment]
This specific PR is rather simple and low-risk.  Do you have any objection on principle on having it integrated?

**Comment by chambart** [comment]
I don't, as long as it is quite clear what are the limitation. Also there might be some reason for removing it later, if for instance the ABI has to change. It should be clear that it might happen.

**Comment by alainfrisch** [comment]
Ok, our flambda gurus seem reluctant to the idea of allowing flambda and non-flambda code to interact (and I can see that this would be difficult for support if flambda changes the ABI).  I don't think it is meaningful to push in this direction, then.

**Pull request closed without merging**